src/v/kafka/server/quota_manager.cc

```diff
@@ -36,55 +58,104 @@ quota_manager::quota_manager()
-        auto full_window = _default_num_windows() * _default_window_width();
-        gc(full_window);
-    });
+    _gc_timer.set_callback([this]() { gc(); });
 }
 
 quota_manager::~quota_manager() { _gc_timer.cancel(); }
 
 ss::future<> quota_manager::stop() {
     _gc_timer.cancel();
-    return ss::make_ready_future<>();
+    co_await _gate.close();
+    co_return co_await ss::make_ready_future<>();
 }
 
 ss::future<> quota_manager::start() {
-    _gc_timer.arm_periodic(_gc_freq);
-    return ss::make_ready_future<>();
+    if (ss::this_shard_id() == _client_quotas.shard_id()) {
+        co_await _client_quotas.reset(client_quotas_map_t{});
+        _gc_timer.arm_periodic(_gc_freq);
+    }
 }
 
-ss::future<quota_manager::client_quotas_t::iterator>
+ss::future<quota_manager::client_quotas_map_t::iterator>
 quota_manager::maybe_add_and_retrieve_quota(
   const std::optional<std::string_view>& quota_id,
   const clock::time_point& now) {
     // requests without a client id are grouped into an anonymous group that
     // shares a default quota. the anonymous group is keyed on empty string.
-    auto qid = quota_id ? *quota_id : "";
-
-    // find or create the throughput tracker for this client
-    //
-    // c++20: heterogeneous lookup for unordered_map can avoid creation of
-    // an sstring here but std::unordered_map::find isn't using an
-    // equal_to<> overload. this is a general issue we'll be looking at. for
-    // now, these client-name strings are small. This will be solved in
-    // c++20 via Hash::transparent_key_equal.
-    auto [it, inserted] = _client_quotas.try_emplace(
-      ss::sstring(qid),
-      client_quota{
-        now,
-        clock::duration(0),
-        {static_cast<size_t>(_default_num_windows()), _default_window_width()},
-        {static_cast<size_t>(_default_num_windows()), _default_window_width()},
-        /// pm_rate is only non-nullopt on the qm home shard
-        (ss::this_shard_id() == quota_manager_shard)
-          ? std::optional<token_bucket_rate_tracker>(
-            {*_target_partition_mutation_quota(),
-             static_cast<uint32_t>(_default_num_windows()),
-             _default_window_width()})
-          : std::optional<token_bucket_rate_tracker>()});
+    auto qid = ss::sstring{quota_id ? *quota_id : ""};
+
+    vassert(_client_quotas, "_client_quotas should have been initialized");
+
+    auto it = _client_quotas->find(qid);
+    auto inserted = false;
+    if (it == _client_quotas->end()) {
+        auto produce_rate = get_client_target_produce_tp_rate(qid);
+        auto fetch_rate = get_client_target_fetch_tp_rate(qid);
+        auto partition_mutation_rate = *_target_partition_mutation_quota();
+        auto default_num_windows = static_cast<uint32_t>(
+          _default_num_windows());
+        auto default_window_width = _default_window_width();
+        auto replenish_threshold = static_cast<uint64_t>(
+          _replenish_threshold().value_or(1));
```

Comment:
I was trying to avoid capturing `this`. But we're already taking a reference to the sharded map as a parameter, so we're already relying on `quota_manager` to outlive this section of the code, so I went ahead and moved these into the `update_func` to simplify.

It's worth noting that moving this into the `update_func` means that we now spend a few more cycles inside the mutex of the `sharded_ptr`. But I expect the map copying to dominate the time inside the `update_func` so this should be negligible.